multiqc/modules/bclconvert/bclconvert.py

```diff
@@ -0,0 +1,696 @@
+                    self.get_bar_data_from_undetermined(self.bclconvert_bylane),
+                    None,
+                    {
+                        "id": "bclconvert_undetermined",
+                        "title": "bclconvert: Undetermined barcodes by lane",
+                        "ylab": "Count",
+                        "tt_percentages": False,
+                        "use_legend": True,
+                        "tt_suffix": "reads",
+                    },
+                ),
+            )
+
+    def _get_genome_size(self):
+        gs = getattr(config, "bclconvert", {}).get("genome_size")
+        if gs:
+            try:
+                gs = float(gs)
+            except ValueError:
+                presets = {"hg19_genome": 2897310462, "hg38_genome": 3049315783, "mm10_genome": 2652783500}
+                if gs in presets:
+                    gs = presets[gs]
+                else:
+                    log.warning(
+                        "The size for genome "
+                        + gs
+                        + " is unknown to MultiQC, "
+                        + "please specify it explicitly or choose one of the following: "
+                        + ", ".join(presets.keys())
+                        + "."
+                    )
+                    gs = None
+        log.debug("Determied genome size as " + str(gs))
+        return gs
+
+    def _set_up_reads_dictionary(self, dict):
+        dict["reads"] = 0
+        dict["yield"] = 0
+        dict["perfect_index_reads"] = 0
+        dict["one_mismatch_index_reads"] = 0
+        dict["basesQ30"] = 0
+        dict["mean_quality"] = 0
+
+    def _parse_single_runinfo_file(self, runinfo_file):
+        # get run id and readlength from RunInfo.xml
+        try:
+            tree = ET.parse(runinfo_file["f"])
+            root = tree.getroot()
+            run_id = root.find("Run").get("Id")
+            read_length = root.find("./Run/Reads/Read[1]").get(
+                "NumCycles"
+            )  # ET indexes first element at 1, so here we're gettign the first NumCycles
+        except:
+            log.error("Could not parse RunInfo.xml to get RunID and read length")
+            raise UserWarning
+        return {"run_id": run_id, "read_length": int(read_length), "cluster_length": int(read_length) * 2}
+
+    def _set_run_details(self):
+        # we might have several runinfo.xmls, but they all ought to be the same run and cluster length
+        for runinfo_file in self.find_log_files("bclconvert/runinfo", filehandles=True):
+            runinfo = self._parse_single_runinfo_file(runinfo_file)
+
+            if self.run_id and runinfo["run_id"] != self.run_id:
+                log.error(
+                    "Input data belongs to multiple sequencer runs (detected multiple run ids). This is not supported by this module: please run MultiQC on each run seperately to generate seperate reports."
+                )
+                raise UserWarning
```

Comment:
This is a bit of a shame - the previous `bcl2fastq` module doesn't have this limitation does it? Is there any way that we can get the module to work with multiple sequencer runs?

One thing is that this effectively means that the CI test don't test the module 😅 But also I wonder if it would be something that people may want to do fairly frequently..?

Reply:
Hey @ewels - thanks for your feedback! We're working with @andrei-seleznev to test this as a replacement for our previous bcl2fastq QC steps and I never even realized it had multi-run support. We normally run MultiQC again over the individual outputs in the rare case that we need to compare between runs (with the option of MegaQC for any systematic comparison).

How does this mess with the CI testing? Wondering if there's a way to get that to work at least.

Reply:
Ah nice! It doesn't break the CI as such, it's just that this module is skipped as there is more than one run in the test data repo. In fairness I guess half of the code in the module is running before it gets to the point of exiting..

Reply:
I believe the current `bcl2fastq` module in the main repo doesn't do undetermined stats recalculation, and this recalculation is the reason for the one-seqrun-only limitation.

The function __recalculate_undetermined()_ in bclconvert.py will recalcualte unknown-read-counds when parsing more than one "bclconvert run" from the same "sequencing run"  (the usecase for this is something like -  take a sequencing run, split the samplesheet by different OverrideCycles settings and run bclconvert on the resulting split, then run MultiQC on all the output - so you have MultiQC parsing "multiple bclconvert runs" from "one sequencing run"). You have to recalcualte the undertermined counts afterwards (which in BclConvert are given _per-lane_) because e.g. some read listed as "undetermined" in one BclConvert-run-outputis actually from another BclConvert-run-output . I think the recalculation of stats doent make sense if the bclconvert runs correspond to different sequencer runs - hence the limitation!

Reply:
Right, makes sense! Sounds like a really nice addition 😀

Can we just disable this calculation if there are multiple runs? Ideally showing the reported undetermined reads if there is one one bclconvert run per sequencing run and hiding them otherwise? Or just skipping undermined completely in these cases.

I can imagine use cases where comparing multiple sequencing runs will be useful for people. For example, having a pool of samples sequenced across multiple runs and trying to compare outputs / pooling and so on.

Reply:
Yeah true, simply disabling the calculation does sound like a good idea! I'll go implement that :)

Reply:
Brilliant - thank you! 👍🏻

Reply:
@andrei-seleznev - do you think that you'll have time to get this written in the next day or two? I'm hoping to get the MultiQC version 1.11 release out this week if possible. If it's too much work then I can merge this PR to get the new module included in the release and we can do a follow-up PR whenever you have time (but if possible I'd prefer to get it all done in one of course).

Cheers,

Phil

Reply:
Yeah I reckon I can get this done in the next day or two, no worries!

Reply:
Awesome 🚀

Reply:
pushed a commit which should resolve this. the module will 'combine' samples across sequencing runs (the user still sees one big table with all the samples) , the per-lane stats are provided per-seqrun-per-lane, and undetermined results are suppressed